keps/sig-node/4960-container-stop-signals/README.md

```diff
@@ -0,0 +1,382 @@
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Container runtimes let you define a STOPSIGNAL to let your container images change which signal is delivered to kill the container. Currently you can only configure this by defining STOPSIGNAL in the container image definition file before you build the image. This becomes difficult to change when you’re using prebuilt images. This KEP proposes to add support to configure custom stop signals for containers from the ContainerSpec. Kubernetes has no equivalent for STOPSIGNAL as part of Pod or Container APIs. This KEP proposes to add support to configure custom stop signals for containers from the ContainerSpec.
+
+## Motivation
+
+Container runtimes like Docker lets you configure the stop signal with which a container would be killed when you start a container. This can be configured either from the container image definition file itself with the STOPSIGNAL instruction or by using the `--stop-signal` flag when starting a container with the respective CLI tool for your runtime. Currently there is no equivalent to this in the Kubernetes APIs.
+
+While managing containers with Kubernetes, if you want to customize an existing image by changing its predefined stop signal or override the default stop signal of SIGTERM, currently you would have to rebuild the container image and update the stop signal at the image definition level. 
+
+Having stop signal as a first class citizen in the container specification would make it easier for users to set custom stop signals for their containers across all types of workloads.
+
+### Goals
+
+- Add a new StopSignal field to ContainerSpec
+- Update the CRI API to take StopSignal as a field in the StopContainerRequest and pass it down to the container runtime via the StopContainer method of remoteRuntimeService
+- Update StopContainer implementation in container runtimes to use the container’s StopSignal, if defined as the stop signal when killing containers.
+
+### Non-Goals
+
+- Change any existing behaviour with how stop signals work when defined in the container image
+
+## Proposal
+
+A new StopSignal field will be added to the Container spec. This new field can take string values, and can be used to define a stop signal when creating Pods.
+
+```go
+// pkg/apis/core/types.go
+type Container struct {
+  // ...
+  // +optional
+  StopSignal string
+}
+
+type ContainerStatus struct {
+  // ...
+  // +optional
+  StopSignal string
```

Comment:
Not sure if this is something we want in Status.

Reply:
Why not?

Reply:
We wouldn't have the stop signal in ContainerStatus once its moved to lifecycle as discussed.

Reply:
See other comments - we absolutlety DO want it in status, I think.

Reply:
I think I meant the "effective" stopsignal -- if the Pod specified one, then that's the one.  If the Pod did not specify, but the image did, then that's the one.  If neither Pod nor image specified, then it's the runtime's default (presumably SIGTERM, but it's not clear if that is part of spec or just convention).

Does that make sense or am I missing something?

Reply:
Yes, sounds good :+1:

Reply:
Makes sense.. yes.

The rest:

SIG<that will be used on stop request> // specified order: pod>image>runtime(NRI?/config/default)
timeout (default 2seconds)
SIG<that will be used on above signal timeout>

given ^ should we add timeout and secondary signal to the status or maybe just the verbose status info map

Reply:
"We already control the timeout with `terminationGracePeriodSeconds`, right?

Is there any configurability for the "final" signal?  It has to be something that results in process termination, and there's a very short list of those uncatchable / unblockable signals.  SIGKILL is pretty much the only option.

Reply:
I was more talking from the CRI implementation side than the kubelet node lifecycle/hook side.  

Container Runtimes receive a stop "Timeout" on the StopContainer request but not for StopPodSandbox.. so when we call stopContainer to flush the containers in the pod we are stuck with using zero for the timeout of each container which results in immediate KILL signals.

The KILL is not configurable at least in containerd and yes is pretty much the only last ditch option.. 

You bring up a great point .. mapping terminationGracePeriodSeconds as a total to StopPodSandbox/StopContainer